Support Wolflink reconnection after unexpected failure

adamkrol93 opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
There was an issue with setting up the whole integration after some internet loss or similar. When wolf gateway is back on-line then theirs system is changing value ids which this integration based on.
Now after some specific error integration will try to connect again and recreate the whole session with fetching new value ids.  
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #44453
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

adamkrol93 committed:
Support reconnection after unexpected fails

ottelo9 commented:
Hey why this fix will not be accepted?

spacegaier commented:
> 
> 
> Hey why this fix will not be accepted?

Because no one has so far reviewed and approved it. In the end it needs an approval from a team member, but also non-team members are encouraged to review, test and document their approval.

spacegaier commented on homeassistant/components/wolflink/__init__.py:
```diff
@@ -65,6 +81,11 @@ async def async_update_data():
+        except ParameterReadError as exception:
+            refetch_parameters = True
+            raise UpdateFailed(
+                "Could not fetch values for parameter. Refreshing value ids."
```
```suggestion
                "Could not fetch values for parameter. Refreshing value IDs."
```

MartinHjelmare commented on homeassistant/components/wolflink/__init__.py:
```diff
@@ -55,8 +56,23 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
+            return {
+                parameter.parameter_id: (parameter.value_id, values[parameter.value_id])
+                for parameter in parameters
+                if parameter.value_id in values.keys()
```
```suggestion
                if parameter.value_id in values
```

MartinHjelmare commented on homeassistant/components/wolflink/sensor.py:
```diff
@@ -64,8 +64,10 @@ def name(self):
         """Return the state. Wolf Client is returning only changed values so we need to store old value here."""
-        if self.wolf_object.value_id in self.coordinator.data:
-            self._state = self.coordinator.data[self.wolf_object.value_id]
+        if self.wolf_object.parameter_id in self.coordinator.data:
```
If the parameter id is not in the fetched data should we still return the old state or return None to represent unknown state?

ottelo9 replied:
No idea, but I think returning the old state is okay.

adamkrol93 replied:
From the perspective of cloud and their (stupid) architecture we should present the old state because they are sending only data which were changed.

MartinHjelmare commented on homeassistant/components/wolflink/__init__.py:
```diff
@@ -105,9 +126,14 @@ async def fetch_parameters(client: WolfClient, gateway_id: int, device_id: int):
+    return [param for param in fetched_parameters if param.name != "Reglertyp"]
+
+
+async def fetch_parameters_init(client: WolfClient, gateway_id: int, device_id: int):
```
We should probably use this function in `async_setup_entry`.

adamkrol93 replied:
Yes, I guess we should.

ottelo9 reviewed (approved):
2021-04-13:
I actually test the component (as a custom component). I'will give a feedback tomorrow.

2021-04-14:
Tested the changed files (includes the requested changes from @MartinHjelmare and @spacegaier) as custom component. The values are no longer updated after a Internet Reconnection (via Router). After a manual "Integration Reload" all values starts updating. The Log says nothing about a error:

`2021-04-14 12:49:16 DEBUG (MainThread) [wolf_smartset.wolf_client] Fetched values: {'LastAccess': '2021-04-14T10:49:16.1880938Z', 'Values': [{'ValueId': 6644838037, 'Value': '38.2', 'State': 1}, {'ValueId': 6644838043, 'Value': '29.7', 'State': 1}, {'ValueId': 6644838046, 'Value': '43', 'State': 1}, {'ValueId': 6644838067, 'Value': '1.72', 'State': 1}, {'ValueId': 6644838070, 'Value': '33', 'State': 1}, {'ValueId': 6644838079, 'Value': '62', 'State': 1}], 'IsNewJobCreated': False}`

.. Internet reconnected ...

`2021-04-14 12:50:16 DEBUG (MainThread) [wolf_smartset.wolf_client] Fetched values: {'LastAccess': '2021-04-14T10:50:16.9472662Z', 'Values': [], 'IsNewJobCreated': False}`

I will test the custom component posted by @adamkrol93 [here](https://github.com/home-assistant/core/issues/44453#issuecomment-759955963). That changed component works at this time.

2021-04-15:
The automatically reload after a Internet Disconnect (via my router) doesnt work!!!

ottelo9 commented:
Please merge 👍

adamkrol93 committed:
Update session at every call. Support Offline devices

adamkrol93 commented:
Ok. I have fixed all change requests plus I have added update session (I do not what it does but they do it in the web application every minute) and added support for Offline devices, some kind of circuit breaker for that.

MartinHjelmare commented on homeassistant/components/wolflink/__init__.py:
```diff
@@ -42,17 +43,40 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
+                raise UpdateFailed(
+                    "Could not fetch values from server because device is Offline."
+                )
+            else:
```
We can remove this `else:` and outdent below since we raise above.

adamkrol93 replied:
raise appeared after else so you are right else was not necessary.

adamkrol93 committed:
Remove unnecessary else branch

MartinHjelmare commented on homeassistant/components/wolflink/__init__.py:
```diff
@@ -61,13 +82,18 @@ async def async_update_data():
             raise UpdateFailed("Invalid authentication during update.") from exception
 
-    coordinator = DataUpdateCoordinator(
+    coordinator: DataUpdateCoordinator = DataUpdateCoordinator(
```
This typing shouldn't be needed.

MartinHjelmare replied:
```suggestion
    coordinator = DataUpdateCoordinator(
```

MartinHjelmare committed:
Clean typing

MartinHjelmare commented on homeassistant/components/wolflink/__init__.py:
```diff
@@ -100,9 +126,14 @@ async def fetch_parameters(client: WolfClient, gateway_id: int, device_id: int):
+    return [param for param in fetched_parameters if param.name != "Reglertyp"]
+
+
+async def fetch_parameters_init(client: WolfClient, gateway_id: int, device_id: int):
```
```suggestion
async def fetch_parameters_init(client: WolfClient, gateway_id: int, device_id: int) -> list[Parameter]:
```
